source/D2Lang/src/D2Unicode/D2UnicodeUtf.cpp

```diff
@@ -0,0 +1,117 @@
+/**
+ * Copyright (c) 2021 Mir Drualga
+ *
+ * Permission is hereby granted, free of charge, to any person
+ * obtaining a copy of this software and associated documentation
+ * files (the "Software"), to deal in the Software without
+ * restriction, including without limitation the rights to use, copy,
+ * modify, merge, publish, distribute, sublicense, and/or sell copies
+ * of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+#include <D2Unicode.h>
+
+#include <stddef.h>
+
+#pragma pack(push, 1)
+
+struct UtfConvertSpec {
+  unsigned short leading_unusable_bits;
+  unsigned short leading_bits;
+  unsigned short num_trailing_code_point_bits;
+  int last_code_point;
+  int first_code_point;
+};
+
+#pragma pack(pop)
+
+// D2Lang.[0x6FC1D768 to 0x6FC1D7D8)
+static const UtfConvertSpec sgUtf8ConvertTable[7] = {
+    { 0x80, 0x00, 0, 0x7F, '\0' },
+    { 0xE0, 0xC0, 6, 0x7FF, 0x80 },
+    { 0xF0, 0xE0, 12, 0xFFFF, 0x800 },
+    { 0xF8, 0xF0, 18, 0x1FFFFF, 0x10000 },
+    { 0xFC, 0xF8, 24, 0x3FFFFFF, 0x200000 },
+    { 0xFE, 0xFC, 30, 0x7FFFFFFF, 0x4000000 },
+    { 0 },
+};
+
+/**
+ * Converts a single UCS-2 character to one or more UTF-8 code units.
+ * Returns the number of UTF-8 code units needed to store the
+ * converted character.
+ *
+ * 1.10: Inline
+ * 1.13D: D2Lang.0x6FC08DC0
+ */
+static int UnicodeCharToUtf8CodePoint(char* dest, unsigned short ch) {
+  int code_point = ch;
+
+  for (size_t table_index = 0;
+      sgUtf8ConvertTable[table_index].leading_unusable_bits != 0;
+      ++table_index) {
+    if (code_point <= sgUtf8ConvertTable[table_index].last_code_point) {
+      unsigned short remaining_bits =
```

Comment:
could you change `unsigned short` and `unsigned char` to `uint16_t` and `uint8_t` ?

Reply:
I have changed the two variables to fixed-width int variants. The struct has also been changed to reflect this. The parameters have not been changed.